rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java

```diff
@@ -1597,28 +1631,22 @@ public void addSignature(BtcECKey federatorPublicKey, List<byte[]> signatures, b
         }
         FederationMember signingFederationMember = federationMember.get();
 
-        BtcTransaction btcTx = provider.getPegoutsWaitingForSignatures().get(new Keccak256(rskTxHash));
-        if (btcTx == null) {
-            logger.warn(
-                "No tx waiting for signature for hash {}. Probably fully signed already.",
-                new Keccak256(rskTxHash)
-            );
-            return;
-        }
-        if (btcTx.getInputs().size() != signatures.size()) {
-            logger.warn(
-                "Expected {} signatures but received {}.",
-                btcTx.getInputs().size(),
-                signatures.size()
-            );
-            return;
+        if (!activations.isActive(ConsensusRule.RSKIP326)) {
+            eventLogger.logAddSignature(signingFederationMember, btcTx, rskTxHashSerialized);
         }
 
-        if (!activations.isActive(ConsensusRule.RSKIP326)) {
-            eventLogger.logAddSignature(signingFederationMember, btcTx, rskTxHash);
+        processSigning(signingFederationMember, signatures, rskTxHashSerialized, btcTx);
+
+        Keccak256 rskTxHash = new Keccak256(rskTxHashSerialized);
+        if (!BridgeUtils.hasEnoughSignatures(btcContext, btcTx) && logger.isDebugEnabled()) {
+            int neededSignatures = federation.getNumberOfSignaturesRequired();
+            logMissingSignatures(btcTx, rskTxHash, neededSignatures, federation.getSize());
+
+            return;
         }
 
-        processSigning(signingFederationMember, signatures, rskTxHash, btcTx, federation);
+        logReleaseBtc(btcTx, rskTxHashSerialized);
+        provider.getPegoutsWaitingForSignatures().remove(rskTxHash);
```

Comment:
I put this only as a side note. We have to be careful with how we are mutating state here. Maybe something to analyze in the future.

Reply:
Agree, it's a nice thing to focus on when we work on refactors next time

Reply:
yes, agree. I'll create a task on the backlog to have it in mind